serialization/rp_ser.c

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <rp_ser.h>
+#include <rp_trans.h>
+#include <rp_os.h>
+#include <rp_errors.h>
+
+#define RP_LOG_MODULE SER_CORE
+#include <rp_log.h>
+
+#define RP_SER_RSP_INITIAL_ARRAY_SIZE 1
+#define RP_SER_CMD_EVT_INITIAL_ARRAY_SIZE 2
+
+static uint8_t endpoint_cnt;
+
+static cmd_handler_t cmd_handler_get(struct rp_ser *rp, uint8_t cmd)
+{
+	cmd_handler_t cmd_handler = NULL;
+
+	for (const struct rp_ser_cmd *iter = rp->decoders->cmd_begin;
+	     iter <  rp->decoders->cmd_end; iter++) {
+		if (cmd == iter->cmd) {
+			cmd_handler = iter->func;
+			break;
+		}
+	}
+
+	return cmd_handler;
+}
+
+static rp_err_t cmd_execute(struct rp_ser *rp, uint8_t cmd, CborValue *it)
+{
+	rp_err_t err;
+	cmd_handler_t cmd_handler;
+
+	cmd_handler = cmd_handler_get(rp, cmd);
+
+	if (cmd_handler) {
+		err = cmd_handler(it);
+	} else {
+		RP_LOG_ERR("Unsupported command received");
+		err = RP_ERROR_NOT_SUPPORTED;
+	}
+
+	return err;
+}
+
+static rp_err_t event_parse(struct rp_ser *rp, uint8_t evt, CborValue *it)
+{
+	rp_err_t err;
+	evt_handler_t evt_handler = NULL;
+
+	for (const struct rp_ser_evt *iter = rp->decoders->evt_begin;
+	     iter <  rp->decoders->evt_end; iter++) {
+		if (evt == iter->evt) {
+			evt_handler = iter->func;
+			break;
+		}
+	}
+
+	if (evt_handler) {
+		err = evt_handler(evt, it);
+	} else {
+		RP_LOG_ERR("Unsupported event received");
+		err = RP_ERROR_NOT_SUPPORTED;
+	}
+
+	return err;
+}
+
+static rp_err_t rp_ser_response_parse(struct rp_ser *rp, CborValue *it)
+{
+	rp_err_t err;
+
+	if (rp->rsp_handler) {
+		err = rp->rsp_handler(it);
+		if (err) {
+			return err;
+		}
+
+		rp->rsp_handler = NULL;
+	}
+
+	return rp_os_response_signal(rp);
+}
+
+static rp_err_t rp_ser_event_parse(struct rp_ser *rp, CborValue *it)
+{
+	uint8_t event;
+
+	if (!cbor_value_is_simple_type(it) ||
+	    cbor_value_get_simple_type(it, &event) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	if (cbor_value_advance_fixed(it) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	return event_parse(rp, event, it);
+}
+
+static rp_err_t rp_ser_cmd_parse(struct rp_ser *rp, CborValue *it)
+{
+	rp_err_t err;
+	uint8_t cmd;
+
+	if (!cbor_value_is_simple_type(it) ||
+	    cbor_value_get_simple_type(it, &cmd) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	if (cbor_value_advance_fixed(it) != CborNoError) {
+		return RP_ERROR_INTERNAL;
+	}
+
+	err = cmd_execute(rp, cmd, it);
+	if (err) {
+		return err;
+	}
+
+	RP_LOG_DBG("Received command 0x%02x", cmd);
+
+	return RP_SUCCESS;
+}
+
+static rp_err_t rp_ser_received_data_parse(struct rp_ser *rp,
+					   const uint8_t *data, size_t len)
+{
+	rp_err_t err;
+	CborParser parser;
+	CborValue value;
+	CborValue recursed;
+	uint8_t packet_type;
+
+	if (cbor_parser_init(data, len, 0, &parser, &value) != CborNoError) {
```

Comment:
There is a trick to initialize cbor parser to allow more values on top level container. Add following line after `cbor_parser_init`.

```
value.remaining = UINT32_MAX;
```

This informs the parser that there are more remaining values after current one and its count is unknown.

Using this approach reduces additional container level and smaller and faster code as the result.

Reply:
I works but I had to also added some NULL value at the end of packet to detect the end of it